### PR TITLE
2nd part distributed stdwells: Write reports only for wells owned.

### PR DESCRIPTION
**Pull request by blattms**
This is done by first examining the local perforations. For this this part is factored out into  a virtual method reportConnections, whose results are later gathered on the process that owns the well.

#2947 is included and needs to be merged first. This one is the next urgent ones to reduce the (already quite high) rebase efforts. There are several people working on the well code.

### CODE REVIEW

**Comment by blattms** [comment]
jenkins build this please

**Comment by blattms** [comment]
jenkins build this please

**Comment by blattms** [comment]
Will post the rest in a 3rd PR once this is merged. Until then I might feel a bit stalled.

File: opm/simulators/wells/WellState.hpp [hunk 1]
```diff
@@ -236,6 +257,33 @@ namespace Opm
...
         WellMapType wellMap_;
 
+        using MPIComm = typename Dune::MPIHelper::MPICommunicator;
+#if DUNE_VERSION_NEWER(DUNE_COMMON, 2, 7)
+        using Communication = Dune::Communication<MPIComm>;
+#else
+        using Communication = Dune::CollectiveCommunication<MPIComm>;
+#endif
+        void gatherVectorsOnRoot(std::vector< data::Connection > from_connections,
+                                 std::vector< data::Connection > to_connections,
```
Comment (bska) [inline opm/simulators/wells/WellState.hpp:272]: Are these supposed to be by value/copies?

Comment (blattms) [reply opm/simulators/wells/WellState.hpp]: Thanks, that was indeed a problem.

**Comment by blattms** [comment]
jenkins build this please

**Comment by blattms** [comment]
rebased. Just hoping nothing went  missing in action.

**Comment by blattms** [comment]
jenkins build this please

**Comment by blattms** [comment]
jenkins build this please

**Comment by blattms** [comment]
rebased and fixed fallout from request for splitting functionality into two functions during review of PR #2947

**Comment by blattms** [comment]
jenkins build this please

File: opm/simulators/wells/BlackoilWellModel_impl.hpp [hunk 2]
```diff
@@ -614,34 +615,42 @@ namespace Opm {
...
+            CheckDistributedWellConnections checker(well, *local_parallel_well_info_[well_index]);
+            bool hasFirstPerforation = false;
+            bool firstOpenCompletion = true;
+
             for (const auto& completion : well.getConnections()) {
+                const int i = completion.getI();
+                const int j = completion.getJ();
+                const int k = completion.getK();
+                const int cart_grid_indx = i + cartDims[0] * (j + cartDims[1] * k);
+                const int active_index = cartesian_to_compressed_[cart_grid_indx];
```
Comment (bska) [inline opm/simulators/wells/BlackoilWellModel_impl.hpp:627]: If you're touching this anyway, now would be a good time to use [`Connection::global_index()`](https://github.com/OPM/opm-common/blob/1ef1646fb191b71a62dbef184942de9d93a21c8c/opm/parser/eclipse/EclipseState/Schedule/Well/Connection.hpp#L108) to save on some arithmetic here:
```C++
const auto active_index =
    this->cartesian_to_compressed_[completion.global_index()];
```

File: opm/simulators/wells/BlackoilWellModel_impl.hpp [hunk 3]
```diff
@@ -839,12 +848,18 @@ namespace Opm {
         // Use the pvtRegionIdx from the top cell
         const auto& perf_data = this->well_perf_data_[wellID];
 
+        // Cater for case where local part might have no perforations.
+        const int pvtreg = perf_data.size() ?
```
Comment (bska) [inline opm/simulators/wells/BlackoilWellModel_impl.hpp:852]: Please write this as `! perf_data.empty()` instead.

File: opm/simulators/wells/WellInterface_impl.hpp [hunk 4]
```diff
@@ -1194,23 +1194,23 @@ namespace Opm
...
+        CheckDistributedWellConnections checker(well_ecl_, parallel_well_info_);
         for (size_t c=0; c<connectionSet.size(); c++) {
             const auto& connection = connectionSet.get(c);
+            const int i = connection.getI();
+            const int j = connection.getJ();
+            const int k = connection.getK();
+
+            const int* cpgdim = cart_dims;
+            const int cart_grid_indx = i + cpgdim[0]*(j + cpgdim[1]*k);
+            const int cell = cartesian_to_compressed[cart_grid_indx];
```
Comment (bska) [inline opm/simulators/wells/WellInterface_impl.hpp:1206]: Same here.  There's no need to perform this arithmetic when the value is already stored in the `Connection` object.

File: opm/simulators/wells/WellState.hpp [hunk 5]
```diff
@@ -319,7 +377,9 @@ namespace Opm
                 : (prod_controls.cmode == Well::ProducerCMode::GRUP);
 
             const double inj_surf_rate = well.isInjector() ? inj_controls.surface_rate : 0.0; // To avoid a "maybe-uninitialized" warning.
-
+            const double local_pressure = well_perf_data_[w].size() ?
```
Comment (bska) [inline opm/simulators/wells/WellState.hpp:380]:
```C++
! well_perf_data_[w].empty()
```

File: tests/test_wellstatefullyimplicitblackoil.cpp [hunk 6]
```diff
@@ -114,16 +117,32 @@ struct Setup
...
+            *ppw = &(*pw);
+            pw->communicateFirstPerforation(true);
+            ++pw;
+            ++ppw;
+        }
+
         state.init(cpress, setup.sched,
-                   setup.sched.getWells(timeStep),
+                   wells, ppinfos,
                    timeStep, nullptr, setup.pu, setup.well_perf_data, setup.st, setup.sched.getWells(timeStep).size());
```
Comment (bska) [inline tests/test_wellstatefullyimplicitblackoil.cpp:146]: I think we can skip the second call to `getWells()` here and just use `wells.size()` instead.

File: opm/simulators/wells/WellState.hpp [hunk 7]
```diff
@@ -225,7 +233,10 @@ namespace Opm
                 if (!this->open_for_output_[well_index])
                     continue;
 
-                auto& well = dw[ itr.first ];
+                const auto& pwinfo = *parallel_well_info_[well_index];
+                using WellT = typename std::remove_reference<decltype(dw[ itr.first ])>::type;
```
Comment (bska) [inline opm/simulators/wells/WellState.hpp:237]: If you use `remove_reference_t<>` here you won't need `typename` or `::type`.

**Review by bska — COMMENTED** [review]
Looks okay for the most part although a few of these sections can be simplified slightly.

**Comment by blattms** [comment]
Comments addressed.

**Comment by blattms** [comment]
jenkins build this please

**Comment by blattms** [comment]
jenkins build this please

**Review by bska — APPROVED** [review]
Thanks for the updates.  This looks good to me.  I'll merge this into master now.

**Comment by blattms** [comment]
Cool, thanks.

**Pull request merged**